compiler/dyno/tools/chpldoc/chpldoc.cpp

```diff
@@ -578,75 +620,81 @@ struct RstSignatureVisitor {
-      if (isPostFixBang) {
-        os_ << "!";
-      } else if (isNilable) {
-        os_ << "?";
+        os_ << "nilable ";
       }
-    } else if (call->isBinaryOp()) {
-      assert(call->numActuals() == 2);
-      outerOp = call->op();
-      if (call->actual(0)->isOpCall()) {
-        innerOp = call->actual(0)->toOpCall()->op();
-        needsParens = needParens(outerOp, innerOp, false, false,
-                                 call->actual(0)->toOpCall()->isUnaryOp(),
-                                 isPostfix(call->actual(0)->toOpCall()), false);
-      }
-      if (needsParens) os_ << "(";
-      call->actual(0)->traverse(*this);
-      if (needsParens) os_ << ")";
-      needsParens = false;
-      bool addSpace = (wantSpaces(call->op(), true) ||
-                       call->op()==USTR("by") ||
-                       call->op()==USTR("align"));
-      if (addSpace)
-        os_ << " ";
-      os_ << call->op();
-      if (addSpace)
-        os_ << " ";
-      if (call->actual(1)->isOpCall()) {
-        innerOp = call->actual(1)->toOpCall()->op();
-        needsParens = needParens(outerOp, innerOp, false,
-                                 false, call->actual(1)->toOpCall()->isUnaryOp(),
-                                 isPostfix(call->actual(1)->toOpCall()) , true);
-        // special case handling things like 3*(4*string)
-        if (!needsParens && innerOp == "*" &&
-            (call->actual(1)->toOpCall()->actual(1)->isTuple() ||
-             call->actual(1)->toOpCall()->actual(1)->isIdentifier())) {
-          needsParens = true;
-        }
-      }
-      if (needsParens) os_ << "(";
-      call->actual(1)->traverse(*this);
-      if (needsParens) os_ << ")";
+    } else {
+      os_ << unaryOp;
+    }
+    opHelper(node, 0, (isPostFixBang || isNilable));
+    if (isPostFixBang || isNilable) {
+      os_ << unaryOp;
+    }
+  }
+
+  void opHelper(const OpCall* node, int pos, bool postfix) {
+    bool needsParens = false;
+    bool isRHS = pos;
+    UniqueString outerOp, innerOp;
+    outerOp = node->op();
+    if (node->actual(pos)->isOpCall()) {
+      innerOp = node->actual(pos)->toOpCall()->op();
+      needsParens = needParens(outerOp, innerOp, node->isUnaryOp(), postfix,
+                               node->actual(pos)->toOpCall()->isUnaryOp(),
+                               isPostfix(node->actual(pos)->toOpCall()), isRHS);
+    }
+    // handle printing parens around tuples
+    // ex: 3*(4*string) != 3*4*string
+    if (isRHS && !needsParens && outerOp == USTR("*") &&
+          ((innerOp == USTR("*") && node->actual(pos)->isOpCall() &&
+            node->actual(pos)->toOpCall()->actual(0)->isIntLiteral() &&
+            (node->actual(pos)->toOpCall()->actual(1)->isTuple() ||
+             node->actual(pos)->toOpCall()->actual(1)->isIdentifier())) ||
+           (node->actual(0)->isIntLiteral() &&
+            (node->actual(pos)->isIdentifier() ||
+             node->actual(pos)->isFnCall())))) {
```

Comment:
I know it's been like this for a while, but I think this would be a good time to split these large conditional expressions into some variables, or perhaps a helper function, so that this conditional here is a bit more readable.

Reply:
yeah, this is super confusing, I refactored it into some helpers